ownCloud/UIKit Extensions/UIImageView+Thumbnails.swift

```diff
@@ -0,0 +1,90 @@
+//
+//  UIImageView+Thumbnails.swift
+//  ownCloud
+//
+//  Created by Michael Neuwert on 31.08.2019.
+//  Copyright © 2019 ownCloud GmbH. All rights reserved.
+//
+
+import UIKit
+import ownCloudSDK
+
+#if canImport(QuickLookThumbnailing)
+import QuickLookThumbnailing
+#endif
+
+extension UIImageView {
+
+	private func cacheThumbnail(image: UIImage, size:CGSize, for item:OCItem, in core:OCCore) {
+
+		guard let itemVersionIdentifier = item.itemVersionIdentifier else { return }
+
+		let specID = item.mimeType != nil ? item.mimeType! : "_none_"
+		let event = OCEvent(type: .retrieveThumbnail,
+							userInfo: [OCEventUserInfoKey(rawValue: "specID") : NSString(string: specID), OCEventUserInfoKey.itemVersionIdentifier : itemVersionIdentifier],
+							ephermalUserInfo: nil,
+							result: nil)
+
+		let thumbnail = OCItemThumbnail()
+		thumbnail.itemVersionIdentifier = item.itemVersionIdentifier
+		thumbnail.maximumSizeInPixels = size
+		thumbnail.mimeType = "image/jpeg"
+		thumbnail.data = image.jpegData(compressionQuality: 1.0)
+		thumbnail.specID = specID
+
+		event.result = thumbnail
+
+		core.handle(event, sender: self)
+	}
+
+	@discardableResult func setThumbnailImage(using core:OCCore, from item:OCItem, with size:CGSize, progressHandler:((_ progress:Progress) -> Void)? = nil) -> Progress? {
+		if item.thumbnailAvailability != .none {
+			let displayThumbnail = { (thumbnail: OCItemThumbnail?) in
+				_ = thumbnail?.requestImage(for: size, scale: 0, withCompletionHandler: { (thumbnail, error, _, image) in
+					if error == nil, image != nil, item.itemVersionIdentifier == thumbnail?.itemVersionIdentifier {
+						OnMainThread {
+							self.image = image
+						}
+					}
+				})
+			}
+
+			weak var weakCore = core
+
+			if let thumbnail = item.thumbnail {
+				displayThumbnail(thumbnail)
+			} else {
+				let activeThumbnailRequestProgress = weakCore?.retrieveThumbnail(for: item, maximumSize: size, scale: 0, retrieveHandler: { (_, _, _, thumbnail, _, progress) in
+					item.thumbnail = thumbnail
+					displayThumbnail(thumbnail)
+
+					// No thumbnail returned by the core, try QuickLook thumbnailing on iOS 13
+					#if canImport(QuickLookThumbnailing)
+					if thumbnail == nil, #available(iOS 13, *) {
+						weakCore?.provideDirectURL(for: item, allowFileURL: true, completionHandler: { (_, url, _) in
+							if let url = url {
+								let thumbnailRequest = QLThumbnailGenerator.Request(fileAt: url, size: size, scale: UIScreen.main.scale, representationTypes: .all)
```

Comment:
@felix-schwarz hm... I think we want icons for few file types... e.g. iWork / MS Office documents...

Reply:
I just noticed another side effect of including the `.icon` type: for f.ex. PDFs, the icon gets stored as thumbnail.

So when the file is downloaded, the icon is still shown - instead of a preview of the PDF that `.lowQualityThumbnail` and `.thumbnail` would have returned (and stored).

.all (incl. .icon) | .thumbnail
-- | --
![Bildschirmfoto 2019-10-16 um 14 23 53](https://user-images.githubusercontent.com/639669/66919703-24e43280-f022-11e9-8ade-54c35c18fd27.png) | ![Bildschirmfoto 2019-10-16 um 14 35 13](https://user-images.githubusercontent.com/639669/66919722-2c0b4080-f022-11e9-9234-ae2626fe4f25.png)

Reply:
> So when the file is downloaded, the icon is still shown - instead of a preview of the (now) local PDF file that `.lowQualityThumbnail` and `.thumbnail` would have returned (and stored).

@felix-schwarz we could make caching of returned thumbnail based on it's type and e.g. if it is `.icon` - don't cache it. WDYT?

Reply:
Yup, it's because the icon got submitted to the cache. Filtering them out would work to avoid this particular problem, but then new problems arise: 

- the QL icon needs to be requested anew for every item
- the (kind of surprisingly long) time that takes, the OC icon is shown - and then replaced by another icon
- this is especially apparent when scrolling

Here's what I see when requesting `.all`, but not saving `.icon`s:

![Icon](https://user-images.githubusercontent.com/639669/66991758-fa00e980-f0c8-11e9-8858-414d051e82cc.gif)

I also observed these updates when downloading a PDF:
1) OC icon
2) QL filetype icon
~ Download ~
3) OC icon
4) QL filetype icon
5) QL thumbnail

![icon2](https://user-images.githubusercontent.com/639669/66992510-3bde5f80-f0ca-11e9-93e5-6aa9edc8a62d.gif)

So I see these options:
### 1) Drop OC icons altogether and rely entirely on QL thumbnails
#### Pro:
- consistent style across all icons, not a mix
- locally generated thumbnails

#### Contra:
- file lists start without icons
- icons start appearing only relatively slow
- icons no longer themeable
- higher memory consumption (OC icon UIImages are cached and reused, QL icons aren't)
- no more any fallback if a client wants to avoid QL thumbnail generation for security reasons

### 2) Mix OC icons and QL icons
#### Pro:
- gaining some additional icons
- locally generated thumbnails

#### Contra:
- file lists start with one icon, which is then swapped out for another icon
- second icon start appearing only relatively slow
- icons no longer themeable
- higher memory consumption (OC icon UIImages are cached and reused, QL icons aren't)

### 3) Use QL only for thumbnails of local files
#### Pro:
- locally generated thumbnails where the server can't provide any
- consistent icons
- icons are themeable
- low memory consumption

#### Con:
- a few file types don't get more specific icons

--- 

I'd go with 3).

If those extra icons are really a concern, I'd suggest adding additional icons for those file types. Even recreating the iOS icons shouldn't take much time: they're mostly "suffix" on a white rectangle (f.ex. for PDF) - or blue rectangles on a white rectangle (f.ex. for RTF).

Reply:
@felix-schwarz I have implemented (3) with latest commit. It is pitty, that `QLThumbnailGenerator` doesn't allow to fetch an icon for **mime-type** and does require file URL necessarily:( Otherwise, option (2) could have been more appealing